karabo_data/geometry2.py

```diff
@@ -0,0 +1,483 @@
+                               + (.5 * self.fs_vec * self.fs_pixels)
+
+    def to_crystfel_geom(self, p, a):
+        name = 'p{}a{}'.format(p, a)
+        c = self.corner_pos
+        return CRYSTFEL_PANEL_TEMPLATE.format(
+            name=name, p=p,
+            min_ss=(a * self.ss_pixels), max_ss=(((a + 1) * self.ss_pixels) - 1),
+            ss_vec=_crystfel_format_vec(self.ss_vec),
+            fs_vec=_crystfel_format_vec(self.fs_vec),
+            corner_x=c[0], corner_y=c[1], coffset=c[2],
+        )
+
+    def snap(self):
+        corner_pos = np.around(self.corner_pos[:2]).astype(np.int32)
+        ss_vec = np.around(self.ss_vec[:2]).astype(np.int32)
+        fs_vec = np.around(self.fs_vec[:2]).astype(np.int32)
+        assert {tuple(np.abs(ss_vec)), tuple(np.abs(fs_vec))} == {(0, 1), (1, 0)}
+        # Convert xy coordinates to yx indexes
+        return GridGeometryFragment(corner_pos[::-1], ss_vec[::-1], fs_vec[::-1])
+
+class GridGeometryFragment:
+    ss_pixels = 64
+    fs_pixels = 128
+
+    # These coordinates are all (y, x), suitable for indexing a numpy array.
+    def __init__(self, corner_pos, ss_vec, fs_vec):
+        self.ss_vec = ss_vec
+        self.fs_vec = fs_vec
+        if fs_vec[0] == 0:
+            # Flip without transposing
+            fs_order = fs_vec[1]
+            ss_order = ss_vec[0]
+            self.transform = lambda arr: arr[..., ::ss_order, ::fs_order]
+            corner_shift = np.array([
+                min(ss_order, 0) * self.ss_pixels,
+                min(fs_order, 0) * self.fs_pixels
+            ])
+            self.pixel_dims = np.array([self.ss_pixels, self.fs_pixels])
+        else:
+            # Transpose and then flip
+            fs_order = fs_vec[0]
+            ss_order = ss_vec[1]
+            self.transform = lambda arr: arr.swapaxes(-1, -2)[..., ::fs_order, ::ss_order]
+            corner_shift = np.array([
+                min(fs_order, 0) * self.fs_pixels,
+                min(ss_order, 0) * self.ss_pixels
+            ])
+            self.pixel_dims = np.array([self.fs_pixels, self.ss_pixels])
+        self.corner_idx = corner_pos + corner_shift
+        self.opp_corner_idx = self.corner_idx + self.pixel_dims
+
+
+class AGIPD_1MGeometry:
+    """Detector layout for AGIPD-1M
+
+    The coordinates used in this class are 3D (x, y, z), and represent multiples
+    of the pixel size.
+    """
+    pixel_size = 2e-7  # 2e-7 metres == 0.2 mm
+    def __init__(self, modules):
+        self.modules = modules  # List of 16 lists of 8 fragments
+
+    @classmethod
+    def from_quad_positions(cls, quad_pos, asic_gap=2, panel_gap=29):
+        """Generate an AGIPD-1M geometry from quadrant positions.
+
+        This produces an idealised geometry, assuming all modules are perfectly
+        flat, aligned and equally spaced within their quadrant.
+
+        The quadrant positions are given in pixel units, referring to the first
```

Comment:
Just out of curiosity how are the quadrant numbers defined? 1 is the upper left corner when looking into the beam? And then going counterclockwise? Which would mean that q1=mod 0-3; q2=mod 4-7; q3=mod 8-11; q4=mod 12-15?

Reply:
yes

Reply:
Yup, the quadrants have to be in the same order as the data to be assembled, and the data is stacked by module number. I'll see if I can clarify this in the docstring.